### PR TITLE
 Compiler: Adds "-Wc++-compat" compiler option

### PR DESCRIPTION
**Pull request by Benichou34**
Adds "-Wc++-compat" compiler option to W1 warning level.
This option warn about ISO C constructs that are outside of the common subset of ISO C and ISO C++.

One of the biggest issue is the size of the empty structure.
An empty struct (or union) has size 0 in C, but size 1 in C++.

For example, in some case struct k_spinlock can be empty. k_spinlock is used in Message Queue Structure (struct k_msgq), and the struct k_msgq doesn't have the same size. That generates an alignment mismatch beetween C and C++.

### CODE REVIEW

File: drivers/serial/uart_stm32.c [hunk 1]
```diff
@@ -311,26 +311,30 @@ static inline u32_t uart_stm32_cfg2ll_hwctrl(enum uart_config_flow_control fc)
  *         LL hardware flow control define.
  * @note   Supports only LL_USART_HWCONTROL_RTS_CTS.
  * @param  fc: LL hardware frlow control definition.
- * @retval UART_CFG_FLOW_CTRL_RTS_CTS, or UART_CFG_PARITY_NONE.
```
Comment (erwango) [inline drivers/serial/uart_stm32.c]: This change seems unrelated

File: drivers/serial/uart_stm32.c [hunk 2]
```diff
@@ -311,26 +311,30 @@ static inline u32_t uart_stm32_cfg2ll_hwctrl(enum uart_config_flow_control fc)
...
+ * @retval UART_CFG_FLOW_CTRL_RTS_CTS, or UART_CFG_FLOW_CTRL_NONE.
  */
 static inline enum uart_config_flow_control uart_stm32_ll2cfg_hwctrl(u32_t fc)
 {
 	if (fc == LL_USART_HWCONTROL_RTS_CTS) {
 		return UART_CFG_FLOW_CTRL_RTS_CTS;
 	}
 
-	return UART_CFG_PARITY_NONE;
```
Comment (erwango) [inline drivers/serial/uart_stm32.c]: Unrelated

**Review by erwango — CHANGES_REQUESTED** [review]
Few bikeshedding comments

File: drivers/serial/uart_stm32.c (hunk 1 above)
Comment (Benichou34) [reply drivers/serial/uart_stm32.c]: I removed this change.

File: drivers/serial/uart_stm32.c (hunk 2 above)
Comment (Benichou34) [reply drivers/serial/uart_stm32.c]: I removed this change.

File: include/spinlock.h [hunk 3]
```diff
@@ -61,6 +61,11 @@ struct k_spinlock {
 	 */
 	size_t thread_cpu;
 #endif
+
+#if !defined(CONFIG_SMP) && !defined(SPIN_VALIDATE)
+	/* Empty struct has size 0 in C, size 1 in C++ */
+	char dummy;
```
Comment (andyross) [inline include/spinlock.h]: No, this isn't right.  We need the optimization provided by the empty struct, otherwise the generated code needs to pass extra pointers around when using spinlocks, when what is desired is that they turn naturally into (argument-free) z_arch_irq_lock() calls.

If you want to remove the use of an empty struct, we need to do more work and remove the spinlock structs from the API arguments (when !SMP) and  enclosing struct declarations entirely.

If you just want C++ code to compile in uniprocessor mode, it's probably enough to put this whole file in a `#ifndef __cplusplus` and replace the handful of functions declared with simple declarations.  The spinlock API is not part of our public API and doesn't need to actually be used from C++ code, you just want it not to break when pulled in as a transitive header, right?

**Review by andyross — CHANGES_REQUESTED** [review]
Sorry, didn't get to this when it was a separate patch.   This trick for spinlocks isn't right, that's performance critical code and the struct really does need to disappear and not have dummy fields.

File: include/spinlock.h (hunk 3 above)
Comment (ceolin) [reply include/spinlock.h]: The __cplusplus guard is not right IMHO, we have to get rid of empty struct. I've a patch that enclosure this struct with ifdef but looked bad to me due the amount of ifdef. I can send it or @andyross if you have an alternative from the top of you head, let me know that I change my patch.
But again, we have to remove empty structs and not workaround it.

Comment (ceolin) [reply include/spinlock.h]: btw, add an element just to satisfy it is *not* right too.

**Review by ceolin — CHANGES_REQUESTED** [review]
add an element just to satisfy  it is not the right fix

File: include/spinlock.h (hunk 3 above)
Comment (Benichou34) [reply include/spinlock.h]: Put the whole file in a #ifndef __cplusplus isn't a valid fix, because the Zephyr kernel can be compiled as C library, and used after by C++ code.

Comment (ceolin) [reply include/spinlock.h]: my fix is not ifdefing the whole file but only this struct, that is why this patch is nasty. This is not only a c++ problem btw, this violates C standard as well.

Comment (Benichou34) [reply include/spinlock.h]: Hello @andyross ,
I take some time to analyze z_arch_irq_lock() calls and I haved looking for the impact of the dummy member in the structure.
With default optimization value "-Os", there is no difference beetween the two version.
Compiler generates exactly the same code for calling z_arch_irq_lock(), in both cases.

So, add a "dummy" field to an empty structure doesn't have any impact to the performances.

**Review by pabigot — CHANGES_REQUESTED** [review]
@Benichou34 Sorry for the delay.  The patch for the empty structure has been superseded by other changes, so should be removed.  Please base this on current master: only the change to the W1 warnings should be left.

I'll approve once the rebase is done, but I note this change isn't particularly useful: standard builds and sanitycheck do not use -DW=1, so the problems with Zephyr builds and C++ (which are many) aren't exposed without special steps.  So this wouldn't have avoided the need to retroactive fixes like #15651.

**Comment by erwango** [comment]
@Benichou34 do you have plans for this PR ?
